pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/text/NativeTextIndexReader.java

```diff
@@ -0,0 +1,141 @@
+
+  private final String _column;
+  private int _numDocs;
+  private final File _indexFile;
+  private final PinotDataBuffer _buffer;
+
+  private FST _fst;
+  private BitmapInvertedIndexReader _invertedIndex;
+
+  public NativeTextIndexReader(String column, File indexDir, int numDocs) {
+    _column = column;
+    _numDocs = numDocs;
+    try {
+      _indexFile = getTextIndexFile(indexDir);
+      _buffer = PinotDataBuffer.loadBigEndianFile(_indexFile);
+
+      populateIndexes();
+    } catch (Exception e) {
+      LOGGER.error("Failed to instantiate Lucene text index reader for column {}, exception {}", column,
+          e.getMessage());
+      throw new RuntimeException(e);
+    }
+  }
+
+  private File getTextIndexFile(File segmentIndexDir) {
+    // will return null if file does not exist
+    File file = SegmentDirectoryPaths.findTextIndexIndexFileNative(segmentIndexDir, _column);
+    if (file == null) {
+      throw new IllegalStateException("Failed to find text index file for column: " + _column);
+    }
+    return file;
+  }
+
+  private void populateIndexes() {
+    long invertedIndexLength = _buffer.getLong(4);
+    long fstDataLength = _buffer.getLong(12);
+    int numBitMaps = _buffer.getInt(20);
+
+    long fstDataStartOffset = NativeTextIndexCreator.HEADER_LENGTH;
+    long fstDataEndOffset = fstDataStartOffset + fstDataLength;
+    ByteBuffer byteBuffer = _buffer.toDirectByteBuffer(fstDataStartOffset, (int) fstDataLength);
+    byte[] arr = new byte[byteBuffer.remaining()];
+    // byteBuffer.get(arr);
+    try {
+      _fst = FST.read(new ByteBufferInputStream(Collections.singletonList(byteBuffer)), ImmutableFST.class, true);
+    } catch (IOException e) {
+      throw new RuntimeException(e.getMessage());
+    }
+
+    long invertedIndexEndOffset = fstDataEndOffset + invertedIndexLength;
+    _invertedIndex =
+        new BitmapInvertedIndexReader(_buffer.view(fstDataEndOffset, invertedIndexEndOffset, ByteOrder.BIG_ENDIAN),
+            numBitMaps);
+  }
+
+  @Override
+  public ImmutableRoaringBitmap getDictIds(String searchQuery) {
+    try {
+      RoaringBitmapWriter<MutableRoaringBitmap> writer = RoaringBitmapWriter.bufferWriter().get();
+      RegexpMatcher.regexMatch(searchQuery, _fst, writer::add);
+      return writer.get();
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception while running query: " + searchQuery, e);
+    }
+  }
+
+  @Override
+  public MutableRoaringBitmap getDocIds(String searchQuery) {
+    try {
+      RoaringBitmapWriter<MutableRoaringBitmap> writer = RoaringBitmapWriter.bufferWriter().get();
+      RegexpMatcher.regexMatch(searchQuery, _fst, writer::add);
+      ImmutableRoaringBitmap matchingDictIds = writer.get();
+      MutableRoaringBitmap matchingDocIds = null;
+
+      for (PrimitiveIterator.OfInt it = matchingDictIds.stream().iterator(); it.hasNext(); ) {
+        int dictId = it.next();
```

Comment:
Streaming a bitmap is a very costly high level API. Use a `PeekableIntIterator` for better results.

Reply:
Will do before commit.